### ISSUE TITLE
What should .type mean on a slice?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#5290]
Here's a semantic question:  What should the following mean?

     var A: [1..10] real;
     var B: (A[1..3].type);

One interpretation would be "B is an array over {1..3] of real."  But another might be "B should be a {1..3} alias of some other array (A?  or some other unspecified array?)."  Or perhaps "this shouldn't be legal."

**Comment by bradcray** [comment]
@vasslitvinov : I have a memory of talking to you about this on the way back from lunch one day when you were wrestling through some challenges that the new array implementation put forth for one of your distributions -- Dimensional I think?  If you have an opinion on this question, would you add it please?

**Comment by vasslitvinov** [comment]
Could you articulate the issue here?

I think in Chapel "var B: something" is defined for any type "something".
If "something" is a type expression, this is also defined because a type expression evaluates to a type.

Is it a new issue on your branch because A[1..3] is no longer a proper array?
If so, let's define a slice at the semantic level.

**Comment by bradcray** [comment]
There's no specific issue on my branch that I'm wrestling with, I'm just trying to capture an interesting question that you asked, the answer to which I'm unsure of.  I think the answer has ramifications for both master and the array views branch.  

The question I was trying to ask was simply "What is B, above?"  Is it completely equivalent to:

    var B: [1..3] real;

or is it something else?  And if something else, what is that something else?

**Comment by vasslitvinov** [comment]
Today A[1..3] is indeed a DefaultRectangular array [1..3] real. So B should be that as well, sharing the domain of A[1..3].

So, yes it is like `var B: [1..3] real;` except the array reuses an existing domain.

There is also an unresolved issue discussed in: #4409 
which does not arise in this simple example.

**Comment by bradcray** [comment]
I agree that that's what happens today.  In this question I was trying to ask "Is this what should happen?"  I.e., is the type of a slice expression simply "the same as an array over the slicing domain with the array's element type" (or perhaps more accurately, "the same as an array over the slicing domain sliced with the array's domain with the array's element type") or is it something different?  If you're comfortable with the current interpretation, that's fine with me...  I was just under the impression that you weren't.  Thanks.

**Comment by vasslitvinov** [comment]
The current interpretation is simple and I am comfortable keeping it that way.

If/when we need to reconsider it, I am open to it as well.

**Comment by bradcray** [comment]
OK, thanks!

**Comment by mppf** [comment]
@bradcray, @vasslitvinov - I think this is more connected to Brad's array view work than the discussion has acknowledged yet.

In particular `A[1..3]` is no longer a DefaultRectangular but instead is an array-view type. Now, we can implement the compiler so that A[1..3].type is the same as the type X from `var X =A[1..3];` but it's not immediately obvious to me why that is more intuitive. I'd expect that making `A[1..3].type` be some sort of array-view type with the domain 1..3.

I think a key question here is whether or not the following 2 things should be the same:

1)
```
return A[1..3].type;
```

2)
```
var B = A[1..3];
return B.type;
```

It might help to answer the question *what is .type really for, anyway?* These use cases come immediately to mind:
 a) Declaring a new variable with the same type
 b) Comparing types e.g. in where clauses

For a), I'm not sure it makes sense to declare new variables of some array-view-slice type. For b), I'm not sure it makes sense to always promote such types to full-on array types. So in other words, the answer still is unclear to me.

**Comment by vasslitvinov** [comment]
So far in Chapel for any type there is a set of values and the default value, so one can declare a variable of that type. If we want to stick with this situation, let's see if we want array views to be a different kind of thing - deserving separate types - than arrays.

For example, [array] views can have runtime types just like arrays. For an array the runtime type component is its domain. For a view the runtime component(s) are the array being aliased and perhaps the domain of the alias. So in the above example `B` would probably behave as an alias for `A[1..3]` even though it it declared with an `=`.

This should guide us whether, at the semantic level, the `.type` of `A[1..3]` and `B` are the same. The answer could be affected by what we discovered in #4409 .

At the same time making a view be an array has its own benefits, like simplicity to the user and ability to use the same code for arrays and for views.

**Comment by benharsh** [comment]
I believe ``var B = A[1..3];`` should definitely result in ``B`` being a new non-view array. If you want to store the alias, use ``ref`` or ``=>``.

I think it would be weird if the following resulted in a alias to some array:
```chpl
var B : (A[1..3]).type;
```

What array? ``A``? Why not just do ``ref B = A[1..3];`` ?

Consider another case:
```chpl
proc foo(arg : [] int) {
  var B : arg.type;
  for b in B do b = 1;
  writeln(B);
}
var A : [1..10] int;
foo(A[1..5]);
writeln(A);
```

I think it would be very surprising behavior if ``A``'s data were modified here. Do we want users to have to check if an argument is a slice/reindex/rankchange? Do we want the casual user to even be aware of the ArrayView types?

I'd advocate that ``(A[1..3]).type`` create a new array, and not an ArrayView.

**Comment by vasslitvinov** [comment]
If an array view is not an array, then we can't pass the `A[1..5]` actual to an `[] int` formal.

If an array view is an array, then certainly ``var B = A[1..3];`` should result in ``B`` being a new array.

**Comment by bradcray** [comment]
Implementation-wise, both on master and my branches, an array view is an array, simply one that has a different domain map.

**Comment by vasslitvinov** [comment]
Does the array descriptor from this domain map reference the array whose view is being taken?

If so, that other array may need to be part of the array-view's runtime type.

**Comment by vasslitvinov** [comment]
In order to merge:

* Brad's "an array view is an array with different domain map" and
* Ben's "`var B = A[1..3]` should make `B` a new non-view array"

we need to support the situation where `var x = y` results in `x` and `y` having different types.

I am less certain at the moment whether `var B: (A[1..3]).type` should make B an array view or be the same as `var B = A[1..3]`. If the former, it would be a view of `A` -- or illegal altogether.

**Comment by mppf** [comment]
I don't think it's a big deal for `var x = y` to allow `x` and `y` to have different types (and in fact the array view branch already does that). (Could a user-defined initializer do that? Perhaps...)

I think it's pretty clear that `var B = A[1..3]` should make a new non-view array.

I'm still not sure what `A[1..3].type` should do and if we don't have any brain-waves about it soon I'd probably favor trying to make it an error in the near term. I think that it's slightly easier to think about if we had `.stype` (for "static type") or equivalent that doesn't get the array's run-time type involved.... but it's not obvious to me what we should be doing even there.

Lastly, I'm personally more interested in making sure that type queries (and other commonly used features) work as expected; if answering this question is not required there , it could be an error for now.

**Comment by vasslitvinov** [comment]
> I think it's pretty clear that var B = A[1..3] should make a new non-view array.

I can relate to that.

> I don't think it's a big deal for var x = y to allow x and y to have different types

Except... how do you spec-ify that? It's like saying that for `var x = y;` when `y` is real, then `x` is integer.

I can see having a special case for array views, however.

**Comment by mppf** [comment]
I ran into the corresponding case for tuples - tuples sometimes have ref fields and sometimes value fields - and when returned they only have value fields. It was "obvious" to me that for tuples, something like this:

``` chapel
proc f(arg) {
  var x:arg.type; // what is the type of x?
}
f( (1,"hi") );  // note the tuple literal and argument have type (1, ref string) to match blank intent
```

Here it seems clear to me that `x` should have a type that does not include the reference, that it should be the same as
``` chapel
var tmp = arg;
var x:tmp.type;
```

**Comment by vasslitvinov** [comment]
What I see here is that the `.type` operation "strips" refs from tuple fields.

My reason this makes sense, perhaps a rephrase of Michael's, is that the user cannot do anything with ref fields. Those can be generated only by the compiler and only as actuals of a method call. (right?)

**Comment by mppf** [comment]
> What I see here is that the .type operation "strips" refs from tuple fields.
>
> My reason this makes sense, perhaps a rephrase of Michael's, is that the user cannot do anything with ref fields. Those can be generated only by the compiler and only as actuals of a method call. (right?)

Right. But, I think there's a case to make that by analogy a slice should also use the non-slice type (as that's what you'd get with a copy or return).

**Comment by bradcray** [comment]
I ran back into this question this week in the following context (thanks to @benharsh for figuring out what was going on):

In our current serialize / deserialize interface, the former acts on a value (the value to serialize) while the latter acts on a type (the type that you're going to create a new instance of when deserializing).  So a pattern like this would work for most types:

```chapel
proc foo(x) {
  var data = x.serialize();
  on Locales[numLocales-1] {
    var copyOfX = (x.type).deserialize(data);
  }
}
```

However, it doesn't work for an array slice:

```chapel
var A: [1..10] real;
foo(A[D]);
```

the reason being that today, `X.type` returns a type describing "an array with type `[D] real` which differs from `X`'s type ("an slice of `A` by `D`").  The result is that we call the array slice's serialize routine and the normal array's deserialize routine.  And there's no reason to believe that these will be compatible with one another (they aren't).

As a side note: Having spent more time away from this issue, it's more obvious to me than it may have been last time around that there's a strong connection between this issue and #8496 (though the comments above do also make this connection).  That is, `X.type` behaves consistently in this context to `var B = X;` (`B`'s type is `X.type`).  This perhaps suggests that the current behavior is reasonable, though I think it can be cause for confusion (as it was for me this week).

**Comment by bradcray** [comment]
Interestingly, while sync vars, like slices, also have the property of resulting in a type change when used to initialize a variable of a given type, `.type` on a sync var returns its actual type rather than the type that such a variable would be.  See the output of this program, for example:

```chapel
var x: sync int;
writeln(x.type:string);
var y: x.type;
writeln(y.type:string);

var a: [1..10] real;
ref b =	a[2..8];
writeln(b.type:string);
writeln(b._instance.type:string);
var c: b.type;
writeln(c.type:string);
writeln(c._instance.type:string);
```

which at present is:

```
_syncvar(int(64))
_syncvar(int(64))
[domain(1,int(64),false)] real(64)
unmanaged ArrayViewSliceArr(real(64),int(64),unmanaged DefaultRectangularDom(1,int(64),false),int(64),unmanaged DefaultRectangularArr(1,int(64),false,real(64),int(64)),bool)
[domain(1,int(64),false)] real(64)
unmanaged DefaultRectangularArr(1,int(64),false,real(64),int(64))
```

This makes me think that perhaps `.type` should always give the _actual_ type of an expression and that we should have variable declarations do whatever type transformations are necessary to turn those types into the type that the variable should be (rather than have `.type` tell white lies as it does today).  Imagine for example that:

```chapel
var x: t = e; 
var y = e2;
```

became:

```chapel
var x: typeDeclToNewVar(t) = e;
var y: untypedDeclToNewVar(e2.type) = e2;
```

where to match today's behavior, typeDeclToNewVar() for `sync int` would return `sync int` and for `array slice` became `array`; and untypedDeclToNewVar() for `sync int` became `int` and for `array slice` became `array`.  Users writing their own types could create such functions to have things like "my integer wrapper" turn into simple integers in the declaration context.

Such an approach seems like it would also help with cases like the one Michael notes above (comparing types in a where clause), as well as with the serialize/deserialize challenge I ran into above, by talking about the actual types of things.  And then any messing about to create a new variable from that type is done only when we're actually creating new variables.

@benharsh, you've probably done the most thinking on this topic.  How does this proposal jibe with your thinking?  And do you know where the white lie about the `.type` on slices comes from today if we wanted to change that?

**Comment by benharsh** [comment]
I think ".type" should return the actual type of the variable. The current state of ".type" on slices is because the compiler replaces that expression with ``chpl__buildArrayRuntimeType(dom: domain, type eltType)``. This little function just returns ``dom.buildArray(eltType)``, and has no regard for what the original ``_array`` type might have been.

I'm still not sure about how we should define this kind of 'unwrapping' initialization. I think there are still the two approaches:
A) user-defined type inference + init= (I think this is what you wrote above)
B) method on RHS that returns a value for the LHS

I find method A a little concerning because it would require that a corresponding 'init=' be available on the inferred type. Do we even know if we want to allow secondary 'init=' implementations in a module that doesn't define the type we're trying to initialize? Or would that be OK if we had C++-style namespaces? Would we suggest that users write ``int.init=(other : MyIntWrapper)`` ? Do we want to use this mechanism for ``in``-intent arguments as well? Anyways, I don't think we need ``typeDeclToNewVar`` in this case, just some function/method to handle the case where the user didn't specify a type.

Method B is most like what we have today with ``chpl__initCopy``, which returns a value. It doesn't rely on an ``init=`` existing either. It would, however, be a break in our variable declaration 'philosophy' which has thus far been presented as 'init='.

Anyways, I think we can change ``.type`` for slices independently of the unwrap-init question (which warrants more discussion).

**Comment by mppf** [comment]
I think this issue is related to #16298 and #16275.
